community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/RawBitsTest.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.schema;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.neo4j.values.Value;
+import org.neo4j.values.Values;
+
+import static java.util.Arrays.asList;
+import static org.junit.Assert.assertEquals;
+
+@RunWith( Parameterized.class )
+public class RawBitsTest
+{
+    @Parameterized.Parameter()
+    public String name;
+
+    @Parameterized.Parameter( 1 )
+    public NumberLayout layout;
+
+    @Parameterized.Parameters( name = "{0}" )
+    public static List<Object[]> layouts()
+    {
+        return asList(
+                new Object[]{"Unique",
+                        new UniqueNumberLayout()
+                },
+                new Object[]{"NonUnique",
+                        new NonUniqueNumberLayout()
+                }
+        );
+    }
+
+    final List<Object> objects = Arrays.asList(
+            Double.NEGATIVE_INFINITY,
+            -Double.MAX_VALUE,
+            Long.MIN_VALUE,
+            Long.MIN_VALUE + 1,
+            Integer.MIN_VALUE,
+            Short.MIN_VALUE,
+            Byte.MIN_VALUE,
+            0,
+            Double.MIN_VALUE,
+            Double.MIN_NORMAL,
+            Float.MIN_VALUE,
+            Float.MIN_NORMAL,
+            1L,
+            1.1d,
+            1.2f,
+            Math.E,
+            Math.PI,
+            (byte) 10,
+            (short) 20,
+            Byte.MAX_VALUE,
+            Short.MAX_VALUE,
+            Integer.MAX_VALUE,
+            9007199254740992D,
+            9007199254740993L,
+            Long.MAX_VALUE,
+            Float.MAX_VALUE,
+            Double.MAX_VALUE,
+            Double.POSITIVE_INFINITY,
+            Double.NaN
+    );
+
+    @Test
+    public void mustSortInSameOrderAsValueComparator() throws Exception
+    {
+        // given
+        List<Value> values = asValueObjects( objects );
+        List<SchemaNumberKey> schemaNumberKeys = asSchemaNumberKeys( values );
+        Collections.shuffle( values );
+        Collections.shuffle( schemaNumberKeys );
+
+        // when
+        values.sort( Values.COMPARATOR );
+        schemaNumberKeys.sort( layout );
+        List<Value> actual = schemaNumberKeys.stream()
+                .map( k -> RawBits.asNumberValue( k.rawValueBits, k.type ) )
+                .collect( Collectors.toList() );
+
+        // then
+        assertSameOrder( actual, values );
```

Comment:
I think we want to test the comparison of all the values to all the values, so N² comparisons. Shuffling and sorting the elements will only give us N·log(N) comparisons. I'd also add in a handful of floats and doubles that only differ by one ulp, e.g. Math.PI vs Math.nextUp/nextDown(Math.PI), and similarly for floats. I'd also add some int/long values that floats/doubles can almost-but-not-quite represent accurately.